vtree/diff.js

```diff
@@ -343,16 +348,23 @@ function reorder(aChildren, bChildren) {
 // TODO: We really want to make this more efficient
 // Finding the current index of any key in the array could be
 // faster.
-function moveItem(array, toIndex, itemKey) {
+function indexOfKey(array, itemKey) {
```

Comment:
i just took a look at what react does and we could achieve O(N) if we were to sacrifice the same things they have given up.

take this example:

```jsx
React.render(
    <ul>
      <li key="1" id="1">1</li>
      <li key="2" id="2">2</li>
      <li key="3" id="3">3</li>
      <li key="0" id="0">0</li>
      <li key="4" id="4">4</li>
  </ul>,
  document.getElementById('example')
);
                                         
setTimeout(function () {
  React.render(
      <ul>
          <li key="0" id="0">0</li>
          <li key="1" id="1">1</li>
          <li key="2" id="2">2</li>
          <li key="3" id="3">3</li>
          <li key="4" id="4">4</li>
      </ul>,
      document.getElementById('example')
  );
}, 500)
```

intuitively there should be one move to get the element with `key="0"` into it's final position (from 3, to 0).  however, react instead produces 3 moves (from 0, to 1; from 1, to 2; from 2, to 3).  moving all these nodes around has some cost associated with it and will trigger blur handlers and break animations on those nodes.  the question is then whether the cost of calculating the efficient moves outweighs the cost (and breaking developer's expectations of nodes remaining in the DOM) of moving lots of nodes in the DOM.  intuitively i believe that for a typical number of keyed nodes (a low number), calculating the efficient patch and keeping more nodes in the DOM more often is the better option but we don't have any conclusive performance metrics to back this up.  http://vdom-benchmark.github.io/vdom-benchmark/ seems to suggest that so far virtual-dom has produced generally significantly better performance than react.

i'd really like to be able to get O(N) working *AND* produce the same set of moves this code currently produces - i know it's possible but for now i'm favoring working code over the most performant code.  when we can get the performant code to be robust then i think we should switch to it.